File: 2023/day07/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// Card strength orders (higher index = stronger)
const cardStrength = "23456789TJQKA"
const cardStrengthJoker = "J23456789TQKA" // J is weakest in Part 2

type Hand struct {
	cards string
	bid   int
}

// getHandType returns the hand type as an integer (higher = stronger)
// 6 = Five of a kind, 5 = Four of a kind, 4 = Full house
// 3 = Three of a kind, 2 = Two pair, 1 = One pair, 0 = High card
func getHandType(hand string) int {
	counts := make(map[rune]int)
	for _, c := range hand {
		counts[c]++
	}

	// Get sorted counts (descending)
	var values []int
	for _, v := range counts {
		values = append(values, v)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(values)))

	// Classify hand type
	switch {
	case len(values) == 1: // Five of a kind [5]
		return 6
	case values[0] == 4: // Four of a kind [4, 1]
		return 5
	case values[0] == 3 && values[1] == 2: // Full house [3, 2]
		return 4
	case values[0] == 3: // Three of a kind [3, 1, 1]
		return 3
	case values[0] == 2 && values[1] == 2: // Two pair [2, 2, 1]
		return 2
	case values[0] == 2: // One pair [2, 1, 1, 1]
		return 1
	default: // High card [1, 1, 1, 1, 1]
		return 0
	}
}

// getHandTypeWithJokers returns hand type with J as wildcards
func getHandTypeWithJokers(hand string) int {
	jokerCount := strings.Count(hand, "J")
	if jokerCount == 0 {
		return getHandType(hand)
	}
	if jokerCount == 5 {
		return 6 // Five of a kind
	}

	// Count non-joker cards
	counts := make(map[rune]int)
	for _, c := range hand {
		if c != 'J' {
			counts[c]++
		}
	}

	// Get sorted counts (descending)
	var values []int
	for _, v := range counts {
		values = append(values, v)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(values)))

	// Add jokers to the highest count
	values[0] += jokerCount

	// Classify hand type
	switch {
	case values[0] == 5: // Five of a kind
		return 6
	case values[0] == 4: // Four of a kind
		return 5
	case values[0] == 3 && len(values) > 1 && values[1] == 2: // Full house
		return 4
	case values[0] == 3: // Three of a kind
		return 3
	case values[0] == 2 && len(values) > 1 && values[1] == 2: // Two pair
		return 2
	case values[0] == 2: // One pair
		return 1
	default: // High card
		return 0
	}
}

// handKey returns a sort key for comparing hands
func handKey(hand string, strengthOrder string) []int {
	key := make([]int, len(hand))
	for i, c := range hand {
		key[i] = strings.IndexRune(strengthOrder, c)
	}
	return key
}

// compareHands compares two hands for sorting (returns true if a < b)
func compareHands(a, b Hand, useJokers bool) bool {
	var typeA, typeB int
	var strengthOrder string

	if useJokers {
		typeA = getHandTypeWithJokers(a.cards)
		typeB = getHandTypeWithJokers(b.cards)
		strengthOrder = cardStrengthJoker
	} else {
		typeA = getHandType(a.cards)
		typeB = getHandType(b.cards)
		strengthOrder = cardStrength
	}

	if typeA != typeB {
		return typeA < typeB
	}

	// Compare card by card
	keyA := handKey(a.cards, strengthOrder)
	keyB := handKey(b.cards, strengthOrder)
	for i := range keyA {
		if keyA[i] != keyB[i] {
			return keyA[i] < keyB[i]
		}
	}
	return false
}

func parseInput(filename string) ([]Hand, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var hands []Hand
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Fields(line)
		bid, _ := strconv.Atoi(parts[1])
		hands = append(hands, Hand{cards: parts[0], bid: bid})
	}
	return hands, scanner.Err()
}

func part1(hands []Hand) int {
	// Make a copy to avoid modifying original
	sorted := make([]Hand, len(hands))
	copy(sorted, hands)

	sort.Slice(sorted, func(i, j int) bool {
		return compareHands(sorted[i], sorted[j], false)
	})

	total := 0
	for rank, hand := range sorted {
		total += (rank + 1) * hand.bid
	}
	return total
}

func part2(hands []Hand) int {
	// Make a copy to avoid modifying original
	sorted := make([]Hand, len(hands))
	copy(sorted, hands)

	sort.Slice(sorted, func(i, j int) bool {
		return compareHands(sorted[i], sorted[j], true)
	})

	total := 0
	for rank, hand := range sorted {
		total += (rank + 1) * hand.bid
	}
	return total
}

func main() {
	// Get the directory of the current executable/source file
	execPath, _ := os.Executable()
	dir := filepath.Dir(execPath)

	// Try multiple possible input locations
	inputPaths := []string{
		filepath.Join(dir, "..", "input.txt"),
		"../input.txt",
		filepath.Join(filepath.Dir(os.Args[0]), "..", "input.txt"),
	}

	var hands []Hand
	var err error

	for _, path := range inputPaths {
		hands, err = parseInput(path)
		if err == nil {
			break
		}
	}

	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(hands))
	fmt.Printf("Part 2: %d\n", part2(hands))
}

File: 2023/day07/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class Solution {

    // Card strength order (higher index = stronger)
    private static final String CARD_STRENGTH = "23456789TJQKA";
    private static final String CARD_STRENGTH_JOKER = "J23456789TQKA";

    public static void main(String[] args) throws IOException {
        String input = Files.readString(Path.of("../input.txt")).trim();
        String[] lines = input.split("\n");

        System.out.println("Part 1: " + part1(lines));
        System.out.println("Part 2: " + part2(lines));
    }

    private static long part1(String[] lines) {
        List<Hand> hands = new ArrayList<>();
        for (String line : lines) {
            String[] parts = line.split("\\s+");
            hands.add(new Hand(parts[0], Integer.parseInt(parts[1])));
        }

        hands.sort(Comparator.comparingInt((Hand h) -> getHandType(h.cards))
                .thenComparing(h -> getCardValues(h.cards, CARD_STRENGTH)));

        long total = 0;
        for (int rank = 1; rank <= hands.size(); rank++) {
            total += (long) rank * hands.get(rank - 1).bid;
        }
        return total;
    }

    private static long part2(String[] lines) {
        List<Hand> hands = new ArrayList<>();
        for (String line : lines) {
            String[] parts = line.split("\\s+");
            hands.add(new Hand(parts[0], Integer.parseInt(parts[1])));
        }

        hands.sort(Comparator.comparingInt((Hand h) -> getHandTypeWithJokers(h.cards))
                .thenComparing(h -> getCardValues(h.cards, CARD_STRENGTH_JOKER)));

        long total = 0;
        for (int rank = 1; rank <= hands.size(); rank++) {
            total += (long) rank * hands.get(rank - 1).bid;
        }
        return total;
    }

    private static int getHandType(String hand) {
        int[] counts = getCardCounts(hand);
        Arrays.sort(counts);
        return classifyHand(counts);
    }

    private static int getHandTypeWithJokers(String hand) {
        int jokerCount = 0;
        for (char c : hand.toCharArray()) {
            if (c == 'J') jokerCount++;
        }

        if (jokerCount == 0) {
            return getHandType(hand);
        }
        if (jokerCount == 5) {
            return 6; // Five of a kind
        }

        // Count non-joker cards
        int[] counts = new int[13];
        for (char c : hand.toCharArray()) {
            if (c != 'J') {
                int idx = CARD_STRENGTH.indexOf(c);
                counts[idx]++;
            }
        }

        Arrays.sort(counts);
        // Add jokers to the highest count
        counts[12] += jokerCount;

        return classifyHand(counts);
    }

    private static int[] getCardCounts(String hand) {
        int[] counts = new int[13];
        for (char c : hand.toCharArray()) {
            int idx = CARD_STRENGTH.indexOf(c);
            counts[idx]++;
        }
        Arrays.sort(counts);
        return counts;
    }

    private static int classifyHand(int[] sortedCounts) {
        // Get the last 5 non-zero counts (sorted ascending)
        int[] relevant = new int[5];
        int idx = 0;
        for (int i = sortedCounts.length - 1; i >= 0 && idx < 5; i--) {
            if (sortedCounts[i] > 0) {
                relevant[idx++] = sortedCounts[i];
            }
        }
        // Sort descending for comparison
        Arrays.sort(relevant);
        reverse(relevant);

        // Classify based on pattern
        if (relevant[0] == 5) return 6;        // Five of a kind
        if (relevant[0] == 4) return 5;        // Four of a kind
        if (relevant[0] == 3 && relevant[1] == 2) return 4;  // Full house
        if (relevant[0] == 3) return 3;        // Three of a kind
        if (relevant[0] == 2 && relevant[1] == 2) return 2;  // Two pair
        if (relevant[0] == 2) return 1;        // One pair
        return 0;                              // High card
    }

    private static void reverse(int[] arr) {
        for (int i = 0; i < arr.length / 2; i++) {
            int temp = arr[i];
            arr[i] = arr[arr.length - 1 - i];
            arr[arr.length - 1 - i] = temp;
        }
    }

    private static CardValues getCardValues(String hand, String cardStrength) {
        int[] values = new int[5];
        for (int i = 0; i < 5; i++) {
            values[i] = cardStrength.indexOf(hand.charAt(i));
        }
        return new CardValues(values);
    }

    private record Hand(String cards, int bid) {}

    private record CardValues(int[] values) implements Comparable<CardValues> {
        @Override
        public int compareTo(CardValues other) {
            for (int i = 0; i < 5; i++) {
                if (this.values[i] != other.values[i]) {
                    return Integer.compare(this.values[i], other.values[i]);
                }
            }
            return 0;
        }
    }
}
